improver/developer_tools/metadata_interpreter.py

```diff
@@ -0,0 +1,563 @@
+                        )
+
+                try:
+                    self.model = MODEL_NAMES[attrs[self.model_id_attr]]
+                except KeyError:
+                    self.errors.append(
+                        f"Attribute {attrs[self.model_id_attr]} is not a valid single "
+                        "model.  If valid for blend, then title attribute is missing "
+                        f"expected substring {BLEND_TITLE_SUBSTR}."
+                    )
+
+    def check_attributes(self, attrs):
+        """Checks for unexpected attributes, then interprets values for model
+        information and checks for self-consistency"""
+        try:
+            permitted_attributes = COMP_ATTRS + DIAG_ATTRS[self.diagnostic]
+        except KeyError:
+            permitted_attributes = COMP_ATTRS.copy()
+
+        if any([attr in NONCOMP_ATTRS for attr in attrs]):
+            self.errors.append(
+                f"Attributes {attrs.keys()} include one or more forbidden "
+                f"values {NONCOMP_ATTRS}"
+            )
+        elif any([attr not in permitted_attributes for attr in attrs]):
+            self.warnings.append(
+                f"{attrs.keys()} include unexpected attributes. Please check the "
+                "standard to ensure this is valid."
+            )
+
+        if self.diagnostic in DIAG_ATTRS:
+            required = DIAG_ATTRS[self.diagnostic]
+            if any([req not in attrs for req in required]):
+                self.errors.append(
+                    f"Attributes {attrs.keys()} missing one or more required "
+                    f"values {required}"
+                )
+
+        if self.field_type != self.ANCIL:
+            if not all([attr in attrs for attr in MANDATORY_ATTRIBUTES]):
+                self.errors.append(
+                    f"Attributes {attrs.keys()} missing one or more mandatory "
+                    f"values {MANDATORY_ATTRIBUTES}"
+                )
+
+            try:
+                self.post_processed = (
+                    True
+                    if PP_TITLE_SUBSTR in attrs["title"]
+                    or BLEND_TITLE_SUBSTR in attrs["title"]
+                    else False
+                )
+            except KeyError:
+                self.errors.append("Cube is missing mandatory title attribute")
+            else:
+                self._check_blend_and_model_attributes(attrs)
+
+    def _check_coords_present(self, coords, expected_coords):
+        """Check whether all expected coordinates are present"""
+        found_coords = [coord for coord in coords if coord in expected_coords]
+        if not set(found_coords) == set(expected_coords):
+            self.errors.append(
+                f"Missing one or more coordinates: found {found_coords}, "
+                f"expected {expected_coords}"
+            )
+
+    def check_spot_data(self, cube, coords):
+        """Check spot coordinates"""
+        self.prod_type = "spot"
+        try:
```

Comment:
Again, I think this is better as an `if "title" in cube.attributes:`. It gets rid of the need to trap an exception.

Reply:
True.  Sorry, I got a bit trigger happy with the "try / except" blocks in this one.